### PR TITLE
[FIX] account_payment_term: allows use of days_next_month in (30,31)

### PR DESCRIPTION
**Pull request by yosa-odoo**
Steps to reproduce:
- Create a new payment term; "days end of month on the 30"
- create an invoice in may
- set the newly created payment term
- create a invoice line
- save

Issue:
the due date (`maturity_date`) is 2024-06-29

Cause:
Use of `relativedelta` with the singular form of `month` -> absolute value

It was set to 2 so it would always give 29 -> year-02-29

and so  `days_next_month = relativedelta(days=min(int(self.days_next_month), (date_end_of_month + relativedelta(month=2)).day))`

would always return 29 since 29 <= (29, 30, 31)

```
>>> import datetime
>>> from dateutil.relativedelta import relativedelta
>>> date_end_of_month
datetime.datetime(2024, 5, 31, 17, 4, 14, 791820)
>>> (date_end_of_month + relativedelta(month=2))
datetime.datetime(2024, 2, 29, 17, 4, 14, 791820)
>>> min(31, 30, (date_end_of_month + relativedelta(month=2)).day)
29
```

Solution:
use the relative form of `relativedelta`; that is, the plural form `months`
```
>>> (date_end_of_month + relativedelta(months=1))
datetime.datetime(2024, 6, 30, 17, 4, 14, 791820)
```
source: https://dateutil.readthedocs.io/en/stable/relativedelta.html

Note:
Had to change the existing test on the overlapping year since
```
>>> december_date
datetime.datetime(2023, 12, 12, 0, 0)
>>> december_date + datetime.timedelta(days=30)
datetime.datetime(2024, 1, 11, 0, 0)
```
and from date the maturity date would be 31st of January

opw-3916451

### CODE REVIEW

**Comment by yosa-odoo** [comment]
The question that remains is:
Do we want the use case of the second part of the test. That is, if days =31 and the following month has < 31 days, the maturity_date not be the last day of the next month but the last day of the month after.

My opinion would be that if we set days=31, we want the last day of the next month (no matter the number of days of the months) and not skip a month

File: addons/account_payment_term/tests/test_payment_term.py
Comment (malb-odoo) [inline addons/account_payment_term/tests/test_payment_term.py:70]: For this we would need the approval of Christophe to be sure IMO

File: addons/account_payment_term/models/account_payment_term.py
Comment (malb-odoo) [inline addons/account_payment_term/models/account_payment_term.py:51]: if we really want that behavior (need to see with po) then a comment would be great

File: addons/account_payment_term/models/account_payment_term.py
Comment (malb-odoo) [inline addons/account_payment_term/models/account_payment_term.py:52]: IMO the variable is not really needed ? If you think it make the code more readable then i'm ok with it

File: addons/account_payment_term/tests/test_payment_term.py
Comment (malb-odoo) [inline addons/account_payment_term/tests/test_payment_term.py:40]: We have two payment term with name special case, maybe we can be more explicit here

File: addons/account_payment_term/tests/test_payment_term.py
Comment (malb-odoo) [inline addons/account_payment_term/tests/test_payment_term.py:89]: already 0 by default in the creation of pay_term_days_end_of_month_30 ?

File: addons/account_payment_term/tests/test_payment_term.py
Comment (malb-odoo) [inline addons/account_payment_term/tests/test_payment_term.py:92]: To be consistent with the docstring what do you think calling this case_1 and for the other case_2 ? and the comment would become useless :smile:

File: addons/account_payment_term/tests/test_payment_term.py
Comment (malb-odoo) [inline addons/account_payment_term/tests/test_payment_term.py:127]: same here, in the docstring it's case 1 and the comments is case 3 then the form is case 30 :exploding_head:, let's use case_1 and case_2 again thanks to the docstring i think it's clear

**Review by malb-odoo — COMMENTED** [review]
Thanks for the fix ! Small comments or suggestions, i would still wait on @chklop insight to be sure about the change :smile:

**Comment by malb-odoo** [comment]
I spoke to Christophe about the change of behavior and he told me that the old behavior should stand. So the test case where it fall to the 29 of february is good. If you have a question about it maybe you can go check with him on thursday :smile:

**Comment by yosa-odoo** [comment]
few examples after discussion with chkl

```
4/06 after=10; days_end_of_month:30 -> 30/07
4/06 after=10; days_end_of_month:29 -> 29/07
4/06 after=10; days_end_of_month:31 -> 31/07
4/06 after=27; days_end_of_month:31 -> 31/08
4/01 after=0; days_end_of_month:31 -> 29/02
12/12 after=30; days_end_of_month:31 -> 29/02
```

**Comment by yosa-odoo** [comment]
Don't approve yet. I have to add a test case

File: addons/account_payment_term/tests/test_payment_term.py
Comment (yosa-odoo) [reply addons/account_payment_term/tests/test_payment_term.py]: yep already at 0, I just put it there so the reader wouldn't need to check the records. Just to make it explicit. But if you want me to remove it I can

**Review by malb-odoo — APPROVED** [review]
LGTM thanks for the fix ! Just a little thing for the assignation of the nb_days but i will not block for that since it's for a readability purpose :smile:

cc: @william-andre

**Review by william-andre — COMMENTED** [review]
Isn't this enough?
```python
        if self.delay_type == 'days_end_of_month_on_the':
            return due_date + relativedelta(days=self.nb_days) + relativedelta(months=1, day=int(self.days_next_month))
```

Also, note that casting to int can cause ugly tracebacks during the onchange (the constraint only checks when saving).
So that probably needs fixing too.
Why is it even a Char field? :fearful: @malb-odoo

**Comment by malb-odoo** [comment]
@william-andre we wanted a field that had a max size of 2 to take a certain size in the list view and if i remember correctly the width on the column were not working :grimacing:

**Comment by william-andre** [comment]
That would be a very bad reason :sweat_smile: 
I just tried it with a Integer, I don't see any issues?

**Comment by malb-odoo** [comment]
> That would be a very bad reason 😅 I just tried it with a Integer, I don't see any issues?

From what i see in the commit message i've putted that This field is a Char because we want the field to be of size 2. Also, we added a constraint that this field must be numeric and between 0 and 31. Now maybe we can make it work with an integer and i was just dumb ahah

**Comment by yosa-odoo** [comment]
For the type of field, I'll let you do what you want in master :)

**Comment by yosa-odoo** [comment]
> [...] the constraint only checks when saving [...] So that probably needs fixing too

What is the flow that you do that is broken?

**Comment by william-andre** [comment]
> What is the flow that you do that is broken?

Have 2 term lines in the config, so that the preview is enabled (not sure why we don't always show the preview...)
One of the lines need to be "days end of month on the 30", and then input any char that is not a valid digit in the field.

File: addons/account_payment_term/models/account_payment_term.py
Comment (william-andre) [inline addons/account_payment_term/models/account_payment_term.py:43]: I think `date_utils` is not used anymore

Comment (yosa-odoo) [reply addons/account_payment_term/models/account_payment_term.py:43]: nice catch

File: addons/account_payment_term/models/account_payment_term.py
Comment (william-andre) [inline addons/account_payment_term/models/account_payment_term.py:42]:
```suggestion
            try:
                days_next_month = int(self.days_next_month)
            except ValueError:
                days_next_month = 1
            return due_date + relativedelta(days=self.nb_days) + relativedelta(months=1, day=days_next_month)
```

**Review by william-andre — APPROVED** [review]
@robodoo r+

**Pull request closed without merging**
